### PR TITLE
Introduced autoVoxelSize

### PR DESCRIPTION
**Pull request by Idclip**

### CODE REVIEW

File: openvdb_points/tools/PointConversion.h [hunk 1]
```diff
@@ -979,6 +1074,151 @@ convertPointDataGridGroup(  Group& group,
...
+
+    for(size_t i = 0; i < 3; ++i) {
+        if(!math::isApproxEqual(minBounds[i], maxBounds[i])) continue;
+        minBounds[i] -= voxelSize / 2.0f;
+        maxBounds[i] += voxelSize / 2.0f;
+    }
+
+    const Vec3d extents(maxBounds - minBounds);
+    double volume = extents[0] * extents[1] * extents[2];
+    assert(!math::isApproxZero(volume));
```
Comment (danrbailey) [inline openvdb_points/tools/PointConversion.h]: It is possible for the size in each axis to just be valid (ie slightly more than 1e-7) resulting in a very tiny resulting volume  (ie 1e-20) which would produce a huge voxel size. Probably worth handling when the volume is also approx zero. (And perhaps unit test points that only just pass this tolerance).

File: openvdb_points/tools/PointConversion.h [hunk 2]
```diff
@@ -979,6 +1074,151 @@ convertPointDataGridGroup(  Group& group,
...
+
+    const size_t estimatedVoxelCount(math::Max((numPoints / pointsPerVoxel), 1ul));
+    voxelSizeFromVolume(volume, estimatedVoxelCount, voxelSize);
+
+    size_t previousSegments(0);
+    size_t currentSegments(1);
+
+    if(interrupter) interrupter->start("autoVoxelSize: Calculating Voxel Size");
```
Comment (danrbailey) [inline openvdb_points/tools/PointConversion.h]: Perhaps just "Calculating automatic voxel size" would be better. Most interrupter messages are in this form.

File: openvdb_points/tools/PointConversion.h [hunk 3]
```diff
@@ -979,6 +1074,151 @@ convertPointDataGridGroup(  Group& group,
...
+
+        MaskGrid::Ptr mask = createGrid<MaskGrid>(false);
+
+        {
+            mask->setTransform(newTransform);
+            tools::PointsToMask<MaskGrid, InterrupterT> pointMaskOp(*mask, interrupter);
+            pointMaskOp.addPoints(positions);
+
+            if(interrupter && util::wasInterrupted(interrupter)) break;
+        }
```
Comment (danrbailey) [inline openvdb_points/tools/PointConversion.h]: Is there a need for this scoping here?

File: openvdb_points/tools/PointConversion.h [hunk 4]
```diff
@@ -979,6 +1074,151 @@ convertPointDataGridGroup(  Group& group,
...
+        const float previousVoxelSize = voxelSize;
+
+        // if the voxel size is invalid (too close to zero) return the previous size
+
+        if(!voxelSizeFromVolume(volume, estimatedVoxelCount, voxelSize)) {
+            voxelSize = previousVoxelSize;
+            break;
+        }
+
+        if(voxelSize / previousVoxelSize > 0.9f) break;
```
Comment (danrbailey) [inline openvdb_points/tools/PointConversion.h]: A comment explaining what the 0.9f means would be good.

File: openvdb_points/tools/PointConversion.h [hunk 5]
```diff
@@ -979,6 +1074,151 @@ convertPointDataGridGroup(  Group& group,
...
+        if(interrupter && util::wasInterrupted(interrupter)) break;
+    }
+
+    if(interrupter) interrupter->end();
+
+    // round the voxel size and return
+
+    return math::Truncate(voxelSize, decimalPlaces);
```
Comment (danrbailey) [inline openvdb_points/tools/PointConversion.h]: Can this be truncated to zero?

File: openvdb_points/unittest/TestPointConversion.cc [hunk 6]
```diff
@@ -551,6 +553,163 @@ TestPointConversion::testStride()
 }
 
 
+////////////////////////////////////////
+
+
+void
+TestPointConversion::testAutoVoxelSize()
```
Comment (danrbailey) [inline openvdb_points/unittest/TestPointConversion.cc]: Looks good. I would suggest adding a few more examples to this: 

* one with points in a very small volume, for example all in 1e-7x1e-7x1e-7 to test the limits of tolerance as mentioned elsewhere
* one with a sparse data set, for example 1000 points in a 1x1x1 cube at the origin, 1000 points in a 1x1x1 cube at y=100, etc
* one with a non-uniform data set, for example 1000 points scattered in a 1x1x1 cube with 1000 points scattered in a 10x10x10x cube with 1000 points scattered in a 100x100x100 cube, etc

File: openvdb_points_houdini/houdini/SOP_OpenVDB_Points.cc [hunk 7]
```diff
@@ -593,11 +593,23 @@ newSopOperator(OP_OperatorTable* table)
         .setDefault("points")
         .setHelpText("Output grid name."));
 
+    parms.add(hutil::ParmFactory(PRM_TOGGLE, "automatic", "Automatic Voxel Size")
+        .setDefault(PRMzeroDefaults)
+        .setHelpText("*BETA* When enabled, automatically attempts to calculate a "
```
Comment (danrbailey) [inline openvdb_points_houdini/houdini/SOP_OpenVDB_Points.cc]: Yep, remove "*BETA*". I'd also suggest phrasing it as calculating the voxel size. If you've provided a transform, it's really just computing the best voxel size for this transform. Suggest we enable it by default as well.

File: openvdb_points_houdini/houdini/SOP_OpenVDB_Points.cc [hunk 8]
```diff
@@ -593,11 +593,23 @@ newSopOperator(OP_OperatorTable* table)
...
+                     "representing the points in a PointDataGrid."));
+
     parms.add(hutil::ParmFactory(PRM_FLT_J, "voxelsize", "Voxel Size")
         .setDefault(PRMpointOneDefaults)
         .setHelpText("The desired voxel size of the new VDB Points grid.")
         .setRange(PRM_RANGE_RESTRICTED, 0, PRM_RANGE_UI, 5));
 
+    parms.add(hutil::ParmFactory(PRM_INT_J, "pointspervoxel", "Points Per Voxel")
+        .setDefault(8)
+        .setHelpText("The average amount of points per voxel to aim for in the new "
```
Comment (danrbailey) [inline openvdb_points_houdini/houdini/SOP_OpenVDB_Points.cc]: "The number of points per voxel to use as the target for automatic voxel size computation." ?

File: openvdb_points_houdini/houdini/SOP_OpenVDB_Points.cc [hunk 9]
```diff
@@ -836,7 +855,32 @@ SOP_OpenVDB_Points::cookMySop(OP_Context& context)
...
+                }
+
+                const openvdb::Vec3R min(bbox.xmin(), bbox.ymin(), bbox.zmin());
+                const openvdb::Vec3R max(bbox.xmax(), bbox.ymax(), bbox.zmax());
+
+                const int pointsPerVoxel = evalInt("pointspervoxel", 0, time);
+
+                hvdbp::HoudiniReadAttribute<openvdb::Vec3R> positions(*(ptGeo->getP()));
+
+                hvdb::Interrupter boss("Points");
```
Comment (danrbailey) [inline openvdb_points_houdini/houdini/SOP_OpenVDB_Points.cc]: Perhaps something slightly more explicit than "Points" ? "OpenVDB Points" ?

File: openvdb_points/tools/PointConversion.h [hunk 10]
```diff
@@ -728,6 +772,57 @@ struct ConvertPointDataGridGroupOp {
     const bool                              mInCoreOnly;
 }; // ConvertPointDataGridGroupOp
 
+inline bool voxelSizeFromVolume(const double volume, const size_t estimatedVoxelCount, float& voxelSize)
```
Comment (danrbailey) [inline openvdb_points/tools/PointConversion.h]: Just so this little function isn't isolated in a completely different place in the code, might be good to include it as a static member of a Local struct nested in the autoVoxelSize function? Lots of example of this pattern in VDB, but here's one - Transform::print(...)

File: openvdb_points/tools/PointConversion.h [hunk 11]
```diff
@@ -979,6 +1074,151 @@ convertPointDataGridGroup(  Group& group,
...
+
+    if(numPoints <= 1) return voxelSize;
+
+    // if the provided bounds are invalid, try to recompute
+
+    if(minBounds[0] > maxBounds[0] ||
+       minBounds[1] > maxBounds[2] ||
+       minBounds[2] > maxBounds[2])
+    {
+        tbb::blocked_range<size_t> range(0, numPoints);
```
Comment (Idclip) [inline openvdb_points/tools/PointConversion.h]: This would be easier to check if minBounds and maxBounds were pointer arguments. That sound good to you?

**Pull request closed without merging**